gh-115988: Add missing ARM64 and RISCV filter in lzma module

ivq opened the pull request:
#115988 

Example:
```python
import lzma

filter = [
    { "id": lzma.FILTER_RISCV },
    { "id": lzma.FILTER_LZMA2 },
]

in_fn = "in.bin"
out_fn = "out.bin"

with open(in_fn, mode="rb") as f:
    in_data = f.read()

with lzma.open(out_fn, mode="wb", filters=filter) as f:
    f.write(in_data)

```

<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--115989.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

ivq committed:
Add missing ARM64 and RISCV filter in lzma module

terryjreedy commented on Doc/library/lzma.rst:
```diff
@@ -349,6 +349,8 @@ options. Valid filter IDs are as follows:
   * :const:`FILTER_ARMTHUMB`
   * :const:`FILTER_POWERPC`
   * :const:`FILTER_SPARC`
+  * :const:`FILTER_ARM64`
```
All these constant names need ! added, as in  `!FILTER_X86`, to suppress the search for a link and the warnings.

ivq replied:
OK. Added.

terryjreedy commented on Modules/_lzmamodule.c:
```diff
@@ -24,6 +24,14 @@
 #endif
 
+
+#ifndef LZMA_FILTER_ARM64
```
I have no idea why the new constants are defined here while the existing ones are not.

ivq replied:
This is for compatibility purpose.
I'm trying to define the constants here to avoid potential build failure.
BCJ filters other than ARM64 and RISCV had been there since the Python lzma module started to exist while the added new filters require lzma v5.4 and v5.6 respectively. If not doing so, building on targets of which lzma version is old will fail.
Also doing so is safe as the lzma library maintains ABI stability: If Python is built with old or new lzma C headers and paired with new runtime lzma dynamic library, everything will just work fine; If Python is built with old or new lzma C headers and paired with old dynamic library, using the new ARM64 or RISCV filter will raise runtime errors while other usage will still work.
Not testing the new added constants in test_lzma is due to the same reason: testing would fail if runtime library is old.

gpshead replied:
Thanks. I added a clarifying comment.

terryjreedy commented on Modules/_lzmamodule.c:
```diff
@@ -372,6 +380,8 @@ lzma_filter_converter(_lzma_state *state, PyObject *spec, void *ptr)
         case LZMA_FILTER_ARM:
         case LZMA_FILTER_ARMTHUMB:
         case LZMA_FILTER_SPARC:
+        case LZMA_FILTER_ARM64:
+        case LZMA_FILTER_RISCV:
```
These additions here and in the next two blocks are consistent with existing code.

terryjreedy reviewed (changes requested):
Does test_lzma test the existing constants in any way?  Such as for validity of the spelling?  If so, the new ones should be added.

Someone more familiar with lzma should do a commit review.

ivq committed:
Suppress doc warning "py:const reference target not found"

serhiy-storchaka reviewed (commented):
They are pretty new features (https://xz.tukaani.org/format/xz-file-format-1.2.0.txt):
```
        1.2.0     2024-01-19    Added RISC-V filter and updated URLs in
                                Sections 0.2 and 7. The URL of this
                                specification was changed.

        1.1.0     2022-12-11    Added ARM64 filter and clarified 32-bit
                                ARM endianness in Section 5.3.2,
                                language improvements in Section 5.4
```
Adding support for them is a new feature in Python, so you need to add versionadded or versionchanged directives in the module documentation and add the corresponding entry in the What's New file.

You should also specify that the new filters (work? are supported? are available?) only with the specified XZ (library? file format?) version. Please add tests for new filters. They should produce expected result or raise an expected exception depending on the xz version.

I am not sure whether these constants should be defined if the underlying library does not support them. The new tests should show what is better.

ivq committed:
Add lzma version

ivq committed:
Add ARM64, RISC-V filter test case

ivq committed:
Update lzma doc

ivq committed:
Add NEWS entry

ivq committed:
Optimise NEWS

serhiy-storchaka commented on Doc/library/lzma.rst:
```diff
@@ -385,6 +396,26 @@ These filters support one option, ``start_offset``. This specifies the address
 that should be mapped to the beginning of the input data. The default is 0.
 
 
+Information about the version of the lzma library in use is available through
```
It is in the "Specifying custom filter chains" subsection. I think that it is better to move it in the "Miscellaneous" section.

ivq replied:
Updated

serhiy-storchaka reviewed (approved):
LGTM. You only need to add an entry in the What's New document.

ivq committed:
Move version doc section

ivq committed:
Add WHATSNEW entry

serhiy-storchaka commented on Doc/library/lzma.rst:
```diff
@@ -324,6 +324,27 @@ Miscellaneous
 
 .. _filter-chain-specs:
```
It should be right before `Specifying custom filter chains`.

ivq replied:
Thank you. Fixed now

ivq committed:
Fix filter chain specs position

ivq committed:
Suppress doc warning

gpshead committed:
Elide upstream binary data from the test.

We only test that the constants are defined if the library we compiled
against was supposed to define them.

Their functionality is up to the external library and not something we
need to test ourselves.

We don't want to accept new binary data from the ill-fated backdoored xz
project into CPython itself.

gpshead committed:
Clarify the documentation a bit.

gpshead committed:
Clarifying comment, docs accuracy.

gpshead commented on Modules/_lzmamodule.c:
```diff
@@ -1591,6 +1611,15 @@ lzma_exec(PyObject *module)
+                     PyUnicode_FromString(LZMA_VERSION_STRING)) < 0) {
+        return -1;
+    }
+    if (PyModule_Add(module, "LZMA_RUNTIME_VERSION",
```
That these are strings is unfortunate, that makes comparisons difficult.  This is seen in test_lzma.py where it has to implement its own version number parsing logic.

All users who want to understand the lzma version numbers are going to need to do the same.

gpshead replied:
I think I'll move that code from test_lzma.py into lzma.py so that these show up as tuples to the user.

gpshead commented on Lib/test/test_lzma.py:
I simplified the test and removed the binary data from the upstream project. It cannot be considered trustworthy anymore: https://www.openwall.com/lists/oss-security/2024/03/29/4

We don't need to test the upstream liblzma BCJ functionality (we never tested any of the existing BCJ filter functionality), just that the new constants exist.

gpshead committed:
Expose the version as a number.

We make lzma_version be the runtime version as that is the one most
interestin to users.  The compile time version remains exposed as
the header version.

gpshead committed:
Merge branch 'main' into xz_bcj_filters

gpshead committed:
minor version explanation tweak.

gpshead committed:
Adjust the whats new text.

gpshead committed:
NEWS wording and formatting tweak.

gpshead committed:
line length

StanFromIreland reviewed (commented):
This has conflicts, and the version has to be changed to `next` as this is not making it to 3.13 :-(

ivq committed:
Merge remote-tracking branch 'origin/main' into xz_bcj_filters

Signed-off-by: Chien Wong <m@xv97.com>

ivq committed:
We've missed the 3.13 window

Signed-off-by: Chien Wong <m@xv97.com>

ivq committed:
Fix accidentally removed newline when merging remote main

Signed-off-by: Chien Wong <m@xv97.com>

ivq committed:
Remove unused import

Signed-off-by: Chien Wong <m@xv97.com>

ivq committed:
Merge remote-tracking branch 'origin/main' into xz_bcj_filters

Signed-off-by: Chien Wong <m@xv97.com>

ivq committed:
Remove lzma version constants

Signed-off-by: Chien Wong <m@xv97.com>

gpshead committed:
versionadded "next" in rst

serhiy-storchaka commented on Doc/whatsnew/3.16.rst:
```diff
@@ -115,6 +115,15 @@ xml
   instead of failing later, when encounter non-ASCII data.
   (Contributed by Serhiy Storchaka in :gh:`62259`.)
 
+lzma
```
Module names should be sorted alphabetically. Move this entry above `os`.

ivq committed:
Put whatsnew entry alphabetically

Signed-off-by: Chien Wong <m@xv97.com>
